src/Language/Haskell/LSP/Core.hs

```diff
@@ -167,872 +272,735 @@ instance E.Exception ProgressCancelledException
+  case res of
+    Left err -> do
+      let msg = T.pack $ unwords
+            ["haskell-lsp:configuration parse error.", show req, show err]
+      sendErrorLog msg
+    Right newConfig ->
+      modifyState $ \ctx -> ctx { resConfig = Just newConfig }
+
+vfsFunc :: (VFS -> b -> (VFS, [String])) -> b -> LspM config ()
+vfsFunc modifyVfs req = do
+  join $ stateState $ \ctx@LanguageContextState{resVFS = VFSData vfs rm} ->
+    let (vfs', ls) = modifyVfs vfs req
+    in (liftIO $ mapM_ (debugM "haskell-lsp.vfsFunc") ls,ctx{ resVFS = VFSData vfs' rm})
+
+-- | Updates the list of workspace folders
+updateWorkspaceFolders :: Message WorkspaceDidChangeWorkspaceFolders -> LspM config ()
+updateWorkspaceFolders (NotificationMessage _ _ params) = do
+  let List toRemove = params ^. J.event . J.removed
+      List toAdd = params ^. J.event . J.added
+      newWfs oldWfs = foldr L.delete oldWfs toRemove <> toAdd
+  modifyState $ \c -> c {resWorkspaceFolders = newWfs $ resWorkspaceFolders c}
 
 -- ---------------------------------------------------------------------
 
 -- | Return the 'VirtualFile' associated with a given 'NormalizedUri', if there is one.
-getVirtualFile :: TVar (LanguageContextData config) -> J.NormalizedUri -> IO (Maybe VirtualFile)
-getVirtualFile tvarDat uri = Map.lookup uri . vfsMap . vfsData . resVFS <$> readTVarIO tvarDat
+getVirtualFile :: NormalizedUri -> LspM config (Maybe VirtualFile)
+getVirtualFile uri = getsState $ Map.lookup uri . vfsMap . vfsData . resVFS
+
+getVirtualFiles :: LspM config VFS
+getVirtualFiles = getsState $ vfsData . resVFS
 
 -- | Dump the current text for a given VFS file to a temporary file,
 -- and return the path to the file.
-persistVirtualFile :: TVar (LanguageContextData config) -> J.NormalizedUri -> IO (Maybe FilePath)
-persistVirtualFile tvarDat uri = join $ atomically $ do
-  st <- readTVar tvarDat
-  let vfs_data = resVFS st
-      cur_vfs = vfsData vfs_data
-      revMap = reverseMap vfs_data
-
-  case persistFileVFS cur_vfs uri of
-    Nothing -> return (return Nothing)
-    Just (fn, write) -> do
-      let revMap' =
-        -- TODO: Does the VFS make sense for URIs which are not files?
-        -- The reverse map should perhaps be (FilePath -> URI)
-            case J.uriToFilePath (J.fromNormalizedUri uri) of
-              Just uri_fp -> Map.insert fn uri_fp revMap
-              Nothing -> revMap
-
-      modifyVFSData tvarDat (\d -> (d { reverseMap = revMap' }, ()))
-      return ((Just fn) <$ write)
+persistVirtualFile :: NormalizedUri -> LspM config (Maybe FilePath)
+persistVirtualFile uri = do
+  join $ stateState $ \ctx@LanguageContextState{resVFS = vfs} ->
+    case persistFileVFS (vfsData vfs) uri of
+      Nothing -> (return Nothing, ctx)
+      Just (fn, write) ->
+        let revMap = case uriToFilePath (fromNormalizedUri uri) of
+              Just uri_fp -> Map.insert fn uri_fp $ reverseMap vfs
+              -- TODO: Does the VFS make sense for URIs which are not files?
```

Comment:
Is this a question for maintainers? Or just a reminder to solve this before merging?